pkg/capr/planner/config.go

```diff
@@ -370,32 +393,59 @@ func addAddresses(secrets corecontrollers.SecretCache, config map[string]interfa
 		}
 
 		driverConfig, err := nodeconfig.ExtractConfigJSON(base64.StdEncoding.EncodeToString(secret.Data["extractedConfig"]))
-		if err != nil || len(driverConfig) == 0 {
+		if err != nil {
 			return fmt.Errorf("error getting machine state JSON for machine %s/%s: %w", entry.Machine.Namespace, entry.Machine.Name, err)
 		}
+		if len(driverConfig) == 0 {
+			return fmt.Errorf("machine state JSON is empty for machine %s/%s", entry.Machine.Namespace, entry.Machine.Name)
+		}
 
 		if !addressProvided {
 			ipAddress = convert.ToString(values.GetValueN(driverConfig, "Driver", "IPAddress"))
 		}
 		if !internalAddressProvided {
 			internalIPAddress = convert.ToString(values.GetValueN(driverConfig, "Driver", "PrivateIPAddress"))
 		}
+		ipv6 = convert.ToString(values.GetValueN(driverConfig, "Driver", "IPv6Address"))
+		driverName = convert.ToString(values.GetValueN(driverConfig, "DriverName"))
 	}
 
-	setNodeExternalIP := ipAddress != "" && internalIPAddress != "" && ipAddress != internalIPAddress
+	// Prefer the public IPv4 address from rancher-machine or the registration command, fall back to the IPv6 address
+	externalIP := firstNonEmpty(ipAddress, ipv6)
+	setNodeExternalIP := externalIP != "" && internalIPAddress != "" && externalIP != internalIPAddress
 
+	// on control-plane nodes with distinct internal/external, advertise the internal
 	if setNodeExternalIP && !isOnlyWorker(entry) {
 		config["advertise-address"] = internalIPAddress
 		config["tls-san"] = append(convert.ToStringSlice(config["tls-san"]), ipAddress)
 	}
 
-	if internalIPAddress != "" {
-		config["node-ip"] = append(convert.ToStringSlice(config["node-ip"]), internalIPAddress)
+	switch driverName {
+	case management.PodDriver:
+		// Skip setting node-ip due to a bug in the pod node driver, which assigns the IPAddress
+		// from the first pod in the namespace instead of the intended target pod.
+		// RKE2/k3s can detect the correct node IP when it is not set.
```

Comment:
Notes to reviewers: Herer is the issue: https://github.com/rancher/rancher/issues/52155